README.md
nodecast
===

nodecast is a high performance live streaming transcoder and stream "muxer"

### transcoder.js
This program is the core of stream-transcoding. It grabs the audio-stream from server.js, playlist.js and the jingle-server and decides, which of them should be put on air now.
The configuration works as follows:

``` javascript
var config = {
    global: {
        version: '0.0.6', // This is the version number of the release. Please do not change it until you update also the script belonging to it
        converterpath: '/usr/bin/ffmpeg' // Specify the path of ffmpeg or avconv here (e.g. /usr/bin/ffmpeg, /usr/local/bin/avconv, etc.)
        convertsamplerate: 48000 // You should set the samplerate of the highest mountpoint defined here
    },
    server: {
        ip: '0.0.0.0', // On which ip should the transcoder listen for dispatchers and statuspage-clients?
        port: 7777, // On which port should the transcoder listen for dispatchers and statuspage-clients?
        allowedips: ['127.0.0.1'] // Which ips should be allowed to stream audio data? Specify the ips of all your dispatcher servers here
    },
    statuspage: {
        allowedips: ['127.0.0.1'], // Which ips should be allowed to display all the statuspages?
        readable: {
            path: '/status', // Specify the url-path of the readable statuspage here
            allowedips: ['0.0.0.0'] // Which ips should be allowed to display the statuspages? Keep in mind that each can see others ip addresses so don't really use 0.0.0.0 here!
        },
        parseable: {
            path: '/status?json', // Specify the url-path of the parseable statuspage here
            allowedips: []
        },
        inspect: {
            path: '/status?inspect', // Specify the url-path of the inspect statuspage here
            allowedips: [],
            options: { // This options will be passed to util.inspect's options parameter
                'showHidden': true,
                'depth': null
            }
        }
    }
};
var mounts = {
    '/main-192.mp3': { // Each mount needs an unique id
        name: 'RaveOne.FM Mainstream 192k', // Specify the icy-name header value here
        url: 'http://raveone.fm/', // Specify the icy-url header value here
        genre: 'RaveOne.FM Mainstream 192k', // Specify the icy-genre header value here
        bitrate: 192, // Conversion bitrate without the 'k' suffix
        samplerate: 48000, // Conversion samplerate
        codec: 'libmp3lame', // Conversion codec
        format: 'mp3', // Conversion format
        metaint: 8192, // Specify the meta-interval here (default: 8192, disable: 0)
        contenttype: 'audio/mpeg', // Specify the content-type header value here
        notices: ['This is', 'a', 'teststream'], // Specify the icy-notice header values here
        allowedips: ['127.0.0.1', '10.135.0.2', '10.135.192.26'] // Which ips should be allowed to stream audio data from only this mountpoint? Specify the ips of all your dispatcher servers here
    },
    '/main-128.mp3': {
        name: 'RaveOne.FM Mainstream 128k',
        url: 'http://raveone.fm/',
        genre: 'RaveOne.FM Mainstream 128k',
        bitrate: 128,
        samplerate: 44100,
        codec: 'libmp3lame',
        format: 'mp3',
        metaint: 8192,
        contenttype: 'audio/mpeg',
        allowedips: ['127.0.0.1', '10.135.0.2', '10.135.192.26']
    }
};
var sources = {
    'dj': { // Each source needs an unique id
        url: 'http://localhost:8080/', // Specify the url of an audio streaming server (e.g. the server.js process) here
        retrywait: 100, // How long to wait after the dj disconnects/times out before reconnecting
        callback: function () {
            source('adjingle', function () {
                source('dj');
            });
        }, // Which functions should be called after the dj disconnects/times out?
        unsyncdiscard: true, // Prevent dispatcher buffer overflow and discard any chunks which do not fit into the unix socketbuffer
        timeout: 1000, // How long to wait before the dj is marked as timed out
        destinations: {
            '/main-192.mp3': {
                priority: 10 // Specify the priority of this source for this destination here
            },
            '/main-128.mp3': {
                priority: 10
            }
        }
    },
    'playlist': {
        url: 'http://127.0.0.1:6666/handsup',
        retrywait: 0,
        unsyncdiscard: true,
        timeout: 1000,
        destinations: {
            '/main-192.mp3': {
                priority: 9
            },
            '/main-128.mp3': {
                priority: 9
            }
        }
    },
    'adjingle': {
        url: 'http://10.135.0.2/streamad/streamad.php',
        isjingle: true, // This tells the setup loop to not autostart this source
        listen: 'SIGUSR2', // Listen for this kill-signal and call this jingle when this event omits
        nativerate: true, // Read this stream at native-rate
        atHours: [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23], // Specify the hours on which the kill-signal should be able to play this jingle
        destinations: {
            '/main-192.mp3': {
                priority: 9
            },
            '/main-128.mp3': {
                priority: 9
            }
        }
    }
};
```

### dispatcher.js
This program takes the stream from one or several transcoder-servers and allows the streams to be mapped to different location paths (e.g. /stream-192k.mp3 and /stream-96k.aac).
The configuration works as follows:

``` javascript
var config = {
    ip: '0.0.0.0', // On which ip should the transcoder listen for dispatchers and statuspage-clients?
    port: 8000, // On which port should the transcoder listen for dispatchers and statuspage-clients?
    maxclients: 10, // Specify the maximum number of overall clients here. If this number of clients are connected, all future clients will be dropped until a slot is free again
    preventclientoverflow: true, // Should the writer discard/drop audio chunks when the client is not responding or reading the buffer anymore?
    prebuffertime: 15000, // How long to prebuffer the data from the stream transcoder to prevent excessive buffering on the client side. When i client connects, this number of seconds will be passed instantly and then the stream will be passed in realtime again.
    servecrossdomainxml: true, // Should the dispatcher serve a crossdomain.xml file? (Used for flash players, embedded in most websites)
    servelistenpls: true, // Should the dispatcher serve a listen.pls file? This file contains all stream-urls until notinlistenpls is set on a specified mount.
    statuspage: {
        allowedips: ['127.0.0.1'], // Which ips should be allowed to display all the statuspages?
        readable: {
            path: '/status', // Specify the url-path of the readable statuspage here
            allowedips: ['0.0.0.0'] // Which ips should be allowed to display the statuspages? Keep in mind that each can see others ip addresses so don't really use 0.0.0.0 here!
        },
        parseable: {
            path: '/status?json',
            allowedips: []
        },
        inspect: {
            path: '/status?inspect', // Specify the url-path of the inspect statuspage here
            allowedips: [],
            options: { // This options will be passed to util.inspect's options parameter
                'showHidden': true,
                'depth': null
            }
        },
        mountslist: {
            path: '/mountslist', // Specify the url-path of the mountslist statuspage here
            allowedips: ['10.135.0.3']
        }
    }
};
var mounts = {
    '/main-192.mp3': { // Each mount needs an unique id
        url: 'http://localhost:7777/main-192.mp3', // Specify the url to the transcoder here (format should be http://transcoderip:transcoderport/transcodermountpoint)
        metaurl: 'http://localhost/meta.php?stream=main-192', // Specify the url to the metadata-server, from where to get the metadata, displayed in the player, here
        maxclients: 10 // Each mountpoint can also have his own maxclients, but not more then the global one.
    },
    '/main-128.mp3': {
        url: 'http://localhost:7777/main-128.mp3',
        metaurl: 'http://localhost/meta.php?stream=main-128',
        maxclients: 10
    },
    '/main-128.aac': {
        url: 'http://localhost:7777/main-128.aac',
        metaurl: 'http://localhost/meta.php?stream=main-aac',
        maxclients: 10
    },
    '/main-64.aac': {
        url: 'http://localhost:7777/main-64.aac',
        metaurl: 'http://localhost/meta.php?stream=main-64.aac',
        maxclients: 10
    }
};
```


ChangeLog
2013.11.22, Version 0.1.2 (Beta)

- transcoder.js
* Removed: -preset ultrafast parameter.. ffmpeg doesn't use it for audio streams
* Bugfix: Callbacks for sources will only be called if the first data event has been omitted

2013.11.22, Version 0.1.1 (Beta)

- transcoder.js
* Combined source and jingle streams into one function and one config (sources) (Still need to change a few things here, don't use jingles in production!)
* Implemented passthrough streams for much better performance
* Other performance increasements and code cleanups (finally... thank god) :-)

- dispatcher.js
* Important bugfix: The metainterval has been calculated wrong: If a chunk was larger than the metainterval bytes set on the transcoder, the chunk hasn't been split into metainterval bytes to fit in there. Instead it has been written to the client without any slicings or splittings done. This caused messed up streams and the listeners had to reconnect to the stream.
* Code cleanups also done here

- functions.js
* Added the functions needed by passthrough streams here

- *
: TODO: Working on a json-config system to prevent overwrites done by updating the programs

2013.11.21, Version 0.0.5 (Beta)

- transcoder.js
* Bugfix~: Default-Codec "libfaac" is an alias to aac but not on all systems out there
* Bugfix~: Default-Retrywait should be really low to prevent underflow in the buffer

- dispatcher.js
* Bugfix: clients now don't get too many listeners caused by a typo in the overflow-check
* Bugfix: process does not crash anymore by serving /listen.pls when mounts aren't setup completely
* Moved init-starts from anywhere between function definitions to the bottom

- functions.js
* Initial commit (moved often called functions from all js files into there)

2013.11.19, Version 0.0.4 (Beta)

- transcoder.js
* Better indentation and syntax
* Added: source callback functions

- dispatcher.js
* Better indentation and syntax

- server.js
* Still a testarea, don't use yet!

- playlist.js
* Better indentation and syntax

2013.11.13, Version 0.0.3 (Alpha)

- transcoder.js
* Added: configuration variable, when to play the jingle at the specified event (if event is emitted AND atHours has the current hour in the list, then play the jingle)
* Removed: status output function completely from code (status output != log output)
* Bugfix: if the mountpoint hasn't been setup completely and a jingle-event occured, this made the process crash and is fixed now
* Bugfix: Weird handling of priority for a source stream not given is fixed now (checking all source and jingle streams for an undefined priority and setting priority to 0)
* Bugfix: Server did not use global configuration variables, fixed now
* Bugfix: Right handling of desynced and unsyncdiscard decision is fixed now

- dispatcher.js

- server.js
* Initial commit

- playlist.js
* Initial commit

2013.10.31, Version 0.0.2 (Alpha)

- transcoder.js
* Renamed: "configs" to "config"
* Renamed: "streams" to "mounts" and "stream()" to "mount()"
* Added: statuspage function from dispatcher.js
* Added: jingle/advertisement function
* Renamed: jingle/advertisement function caller from "adafterdisconnect" to "jingleafterdisconnect"
* Renamed: "adsources" to "jingles"
* Bugfix: mounts[mountpoint].analyzeduration wasn't passed correctly
* Changed: the uniqid-generation behavior from "uniqid('', true)" to "sourcename+'_'+uniqid('', true)"
* Removed: several log() calls, because we don't need debugging anymore
* Bugfix: http-server now checks, whetever the request method is GET and otherwise drops the connection instantly
* Removed: old comented out test space

- dispatcher.js
* Removed: useless 'drain' listener (Usually for testing out things)
* Added: mounts parameter: mounts[mountname].notinlistenpls with witch you can decide whetever this mount should not be written into the stream.pls
* Bugfix: makemeta will now be called by the clienting function with the check-interval specified by the transcoder's metainterval
* Bugfix: makemeta will now be called by an interval and only requests and converts the metaurl data to meta information, when icy-metaint is set correct
* Bugfix: icy-metaint header has been passed with type 'string' instead of 'number'. Now calling parseInt() on it
* Removed: old comented out test space

2013.10.30, Version 0.0.1 (preAlpha)

- transcoder.js
* Initial commit

- dispatcher.js
* Initial commit


userapi.php
<?php

$users = array(
    "root" => array(
        "password" => "4d2ece30e1a503e92939b3bbbc9c96d6",
        "methods" => array(
            "*"
        )
    ),
    "dj" => array(
        "password" => "7d04e57900e1568abb18782fb70c4a28",
        "methods" => array(
            "shoutcast",
            "icecast"
        ),
        "cond" => (date("G") >= 12 && date("G") <= 14)
    )
);

if(!isset($_GET["authstring"]) || empty($_GET["authstring"]) || !isset($_GET["method"]) || empty($_GET["method"])) {
    die("0"); // If no authstring or no method is given, exit with allowness 0
}

list($username, $password) = explode(":", $_GET["authstring"]); // Split authstring into two parts: username and password
$password = md5($password); // Hash password with md5 (this is insecure but we don't need better passwords here)

if(!isset($users[$username]) || empty($users[$username]) || $users[$username]["password"] !== $password) {
    die("0"); // If username does not exist or password is wrong, exit with alloweness 0
}

if(!in_array($_GET["method"], $users[$username]["methods"]) && !in_array('*', $users[$username]["methods"])) {
    die("0"); // If this user is not allowed to call the server with the called methot, exit with allowness 0
}

if(isset($users[$username]["cond"]) && !isset($users[$username]["cond"])) {
    die("2"); // If a condition is set and the condition is met, exit with allowness 2, which means, allow to kick current dj and stream
}

die("1"); // If none of the above conditions were met, exit wich allowness 1, which means, allow to connect to stream


dispatcher.js
#!/usr/bin/env node

'use strict';
/*jslint
   browser: false, devel: false, node: true, rhino: false, passfail: false,
   bitwise: false, debug: false, eqeq: false, evil: false, forin: false,
   newcap: false, nomen: true, plusplus: true, regexp: false, unparam: false,
   sloppy: false, stupid: false, sub: false, vars: false, white: false,
   indent: 4, maxlen: 256
*/

// Configuration (Modify these to your needs)
var config = {};
var mounts = {};

var functions = require('./functions.js');
var http = functions.http;
var in_array = functions.in_array;
var os = functions.os;
var icystring = functions.icystring;
var util = functions.util;
var log = functions.log;
var uniqid = functions.uniqid;

var makemeta = function (metadata) {
    if (typeof metadata === 'string') {
        metadata = {StreamTitle: metadata};
    } else if (!metadata || typeof metadata === 'object') {
        return '';
    }
    var string = icystring(metadata),
        length = Buffer.byteLength(string),
        buflen = Math.ceil(length / 16),
        buffer = new Buffer(buflen * 16 + 1),
        written = buffer.write(string, 1);
    buffer[0] = buflen;
    buffer.fill(0, written + 1);
    return buffer;
};

var server = http.createServer(function (req, res) {
    if (req.method.toUpperCase() !== 'GET') {
        log(req.socket.remoteAddress + ':' + req.socket.remotePort + ' tried method ' + req.method.toUpperCase());
        res.write('<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">' +
                  '<html>' +
                  '<head>' +
                  '<title>405 Method Not Allowed</title>' +
                  '</head>' +
                  '<body>' +
                  '<h1>Method Not Allowed</h1>' +
                  '<p>The requested method ' +
                  req.method.toUpperCase() +
                  'is not allowed on this server.</p>' +
                  '<hr>' +
                  '<address>nodecast Server at ' +
                  req.socket.localAddress +
                  ' Port ' +
                  req.socket.localPort +
                  '</address>' +
                  '</body>' +
                  '</html>');
    } else if (req.url === '/crossdomain.xml' && config.servecrossdomainxml) {
        res.writeHead(200, {
            'content-type': 'text/xml',
            'connection': 'close'
        });
        res.end('<?xml version="1.0"?>' +
                '<!DOCTYPE cross-domain-policy SYSTEM "http://www.macromedia.com/xml/dtds/cross-domain-policy.dtd">' +
                '<cross-domain-policy>' +
                '<allow-access-from domain="*" to-ports="*" />' +
                '</cross-domain-policy>');
    } else if (req.url === '/listen.pls' && config.servelistenpls) {
        res.writeHead(200, {
            'content-type': 'audio/x-scpls',
            'connection': 'close'
        });
        res.write('[playlist]\n');
        res.write('NumberOfEntries=' + Object.keys(mounts).length + '\n\n');
        var filenum = 0;
        Object.keys(mounts).forEach(function (mountpoint) {
            if (!mounts[mountpoint].notinlistenpls && mounts[mountpoint]._) {
                filenum++;
                res.write('File' + filenum + '=http://' + req.headers.host + mountpoint + '\n');
                res.write('Title' + filenum + '=' + mounts[mountpoint]._.headers['icy-name'] + '\n');
                res.write('Length' + filenum + '=-1\n');
            }
        });
        res.end();
    } else if (config.statuspage &&
               config.statuspage.readable &&
               config.statuspage.readable.path &&
               req.url === config.statuspage.readable.path &&
               ((in_array('*', config.statuspage.readable.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.readable.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.readable.allowedips)) ||
                (in_array('*', config.statuspage.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.allowedips)))) {
        res.writeHead(200, {
            'content-type': 'text/plain',
            'connection': 'close'
        });
        var uptime = process.uptime(),
            listener = 0,
            prebuffersize = 0,
            bytesrcvd = 0,
            bytessent = 0,
            systemload = Math.round(os.loadavg()[0] * 100) + '% (' + os.loadavg().join(' ') + ')',
            memoryheap = process.memoryUsage();
        memoryheap = Math.round(memoryheap.heapUsed / memoryheap.heapTotal * 100) + '%';
        Object.keys(mounts).forEach(function (mountpoint) {
            listener += Object.keys(mounts[mountpoint]._.clients).length;
            bytesrcvd += mounts[mountpoint]._.bytesrcvd;
            bytessent += mounts[mountpoint]._.bytessent;
            Object.keys(mounts[mountpoint]._.prebuffers).forEach(function (prebufferkey) {
                prebuffersize += mounts[mountpoint]._.prebuffers[prebufferkey].length;
            });
        });
        res.write('Uptime: ' + uptime + '\n' +
                  'Listener: ' + listener + '\n' +
                  'Bytes received: ' + bytesrcvd + '\n' +
                  'Bytes sent: ' + bytessent + '\n' +
                  'System load: ' + systemload + '\n' +
                  'Memory heap: ' + memoryheap + '\n' +
                  'Prebuffer size: ' + prebuffersize + '\n');
        Object.keys(mounts).forEach(function (mountpoint) {
            res.write('Mount ' + mountpoint + '\n');
            Object.keys(mounts[mountpoint]._.prebuffers).forEach(function (prebufferkey) {
                prebuffersize += mounts[mountpoint]._.prebuffers[prebufferkey].length;
            });
            res.write(' Prebuffer ' + prebuffersize + ' (' + Object.keys(mounts[mountpoint]._.prebuffers).length + ')\n');
            Object.keys(mounts[mountpoint]._.clients).forEach(function (clientid) {
                res.write(' Client ' + clientid + ' ' + mounts[mountpoint]._.clients[clientid].req.socket.remoteAddress + ':' + mounts[mountpoint]._.clients[clientid].req.socket.remotePort + ' ' + mounts[mountpoint]._.clients[clientid].status.sent + ' ' +
                          mounts[mountpoint]._.clients[clientid].status.overflowed + '\n');
            });
        });
        res.end();
    } else if (config.statuspage &&
               config.statuspage.parseable &&
               config.statuspage.parseable.path &&
               req.url === config.statuspage.parseable.path &&
               ((in_array('*', config.statuspage.parseable.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.parseable.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.parseable.allowedips)) ||
                (in_array('*', config.statuspage.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.allowedips)))) {
        res.writeHead(200, {
            'content-type': 'application/json',
            'connection': 'close'
        });
        res.end(JSON.stringify({'config': config, 'mounts': mounts}));
    } else if (config.statuspage &&
               config.statuspage.inspect &&
               config.statuspage.inspect.path &&
               req.url === config.statuspage.inspect.path &&
               ((in_array('*', config.statuspage.inspect.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.inspect.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.inspect.allowedips)) ||
                (in_array('*', config.statuspage.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.allowedips)))) {
        res.writeHead(200, {
            'content-type': 'application/json',
            'connection': 'close'
        });
        res.end(util.inspect({'config': config, 'mounts': mounts}, config.statuspage.inspect.options));
    } else if (config.statuspage &&
               config.statuspage.mountslist &&
               config.statuspage.mountslist.path &&
               req.url === config.statuspage.mountslist.path &&
               ((in_array('*', config.statuspage.mountslist.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.mountslist.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.mountslist.allowedips)) ||
                (in_array('*', config.statuspage.allowedips) ||
                 in_array('0.0.0.0', config.statuspage.allowedips) ||
                 in_array(req.socket.remoteAddress, config.statuspage.allowedips)))) {
        res.writeHead(200, {
            'content-type': 'application/json',
            'connection': 'close'
        });
        var mountslist = [];
        Object.keys(mounts).forEach(function (mountname) {
            mountslist.push(mountname);
        });
        res.end(JSON.stringify(mountslist));
    } else if (mounts[req.url] && mounts[req.url]._) {
        var sumclients = 0,
            clientrealheaders = mounts[req.url]._.headers,
            clientmetaint = 0,
            clientid = uniqid('', true),
            clientstatus = {};
        Object.keys(mounts).forEach(function (mountpoint) {
            if (mounts[mountpoint]._ && mounts[mountpoint]._.clients) {
                sumclients += Object.keys(mounts[mountpoint]._.clients).length;
            }
        });
        if ((!config.maxclients ||
             config.maxclients > sumclients) &&
            (!mounts[req.url].maxclients ||
             mounts[req.url].maxclients > Object.keys(mounts[req.url]._.clients).length)) {
            
            if (req.headers['icy-metadata']) {
                clientmetaint = clientrealheaders['icy-metaint'];
            } else {
                clientmetaint = 0;
                delete clientrealheaders['icy-metaint'];
            }
            res.sendDate = false;
            res.writeHead(200, clientrealheaders);
            clientstatus = {'overflowed': false, 'sent': 0, 'metaint': clientmetaint, 'metaintcycle': 0};
            var resprebuffers = mounts[req.url]._.prebuffers;
            Object.keys(resprebuffers).forEach(function (resprebufferkey) {
                var chunk = resprebuffers[resprebufferkey];
                if (clientstatus.metaint && clientstatus.metaintcycle + chunk.length > clientstatus.metaint) {
                    var remainchunk = new Buffer(chunk.length);
                    chunk.copy(remainchunk);
                    while (clientstatus.metaintcycle + remainchunk.length > clientstatus.metaint) {
                        res.write(remainchunk.slice(0, clientstatus.metaint - clientstatus.metaintcycle));
                        res.write(mounts[req.url]._.meta);
                        remainchunk = remainchunk.slice(clientstatus.metaint - clientstatus.metaintcycle, chunk.length);
                        clientstatus.metaintcycle = 0;
                        clientstatus.sent += mounts[req.url]._.meta.length;
                        mounts[req.url]._.bytessent += mounts[req.url]._.meta.length;
                    }
                    if (remainchunk.length > 0) {
                        res.write(remainchunk);
                        clientstatus.metaintcycle = remainchunk.length;
                    }
                } else {
                    res.write(chunk);
                    clientstatus.metaintcycle += chunk.length;
                }
                clientstatus.sent += chunk.length;
                mounts[req.url]._.bytessent += chunk.length;
            });
            mounts[req.url]._.clients[clientid] = {'status': clientstatus, 'req': req, 'res': res};
            res.once('close', function () {
                delete mounts[req.url]._.clients[clientid];
            });
        } else {
            res.end('No slot available');
        }
    } else {
        res.end('<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">' +
                '<html>' +
                '<head>' +
                '<title>404 Not Found</title>' +
                '</head>' +
                '<body>' +
                '<h1>Not Found</h1>' +
                '<p>The requested URL ' +
                req.url +
                ' was not found on this server.</p>' +
                '<hr>' +
                '<address>nodecast Server at ' +
                req.socket.localAddress +
                ' Port ' +
                req.socket.localPort +
                '</address>' +
                '</body>' +
                '</html>');
    }
});

var clienting = function (mountpoint) {
    log('Spawned clienting(' + mountpoint + ')');
    if (!mounts[mountpoint].prebuffertime) {
        if (!config.prebuffertime) {
            mounts[mountpoint].prebuffertime = false;
        } else {
            mounts[mountpoint].prebuffertime = config.prebuffertime;
        }
    }
    if (mounts[mountpoint].prebuffertime === 0) {
        mounts[mountpoint].prebuffertime = false;
    }
    http.get(mounts[mountpoint].url, function (res) {
        if (!mounts[mountpoint]._) {
            mounts[mountpoint]._ = {};
            mounts[mountpoint]._.headers = {};
            mounts[mountpoint]._.clients = {};
            mounts[mountpoint]._.prebuffers = [];
            mounts[mountpoint]._.bytesrcvd = 0;
            mounts[mountpoint]._.bytessent = 0;
            mounts[mountpoint]._.meta = new Buffer([0]);
        }
        mounts[mountpoint]._.headers = res.headers;
        Object.keys(mounts[mountpoint]._.headers).forEach(function (headerkey) {
            if (typeof mounts[mountpoint]._.headers[headerkey] === 'string' &&
                parseInt(mounts[mountpoint]._.headers[headerkey], 0) + '' === mounts[mountpoint]._.headers[headerkey] &&
                parseInt(mounts[mountpoint]._.headers[headerkey], 0) + 1 - 1 === parseInt(mounts[mountpoint]._.headers[headerkey], 0)) {
                mounts[mountpoint]._.headers[headerkey] = parseInt(mounts[mountpoint]._.headers[headerkey], 0);
            }
        });
        res.on('data', function (chunk) {
            Object.keys(mounts[mountpoint]._.clients).forEach(function (clientid) {
                if (mounts[mountpoint]._.clients[clientid].res.writable && !(mounts[mountpoint]._.clients[clientid].res.socket._writableState.length !== 0 && mounts[mountpoint].preventclientoverflow)) {
                    if (mounts[mountpoint]._.clients[clientid].status.metaint && mounts[mountpoint]._.clients[clientid].status.metaintcycle + chunk.length > mounts[mountpoint]._.clients[clientid].status.metaint) {
                        var remainchunk = new Buffer(chunk.length);
                        chunk.copy(remainchunk);
                        while (mounts[mountpoint]._.clients[clientid].status.metaintcycle + remainchunk.length > mounts[mountpoint]._.clients[clientid].status.metaint) {
                            mounts[mountpoint]._.clients[clientid].res.write(remainchunk.slice(0, mounts[mountpoint]._.clients[clientid].status.metaint - mounts[mountpoint]._.clients[clientid].status.metaintcycle));
                            mounts[mountpoint]._.clients[clientid].res.write(mounts[mountpoint]._.meta);
                            remainchunk = remainchunk.slice(mounts[mountpoint]._.clients[clientid].status.metaint - mounts[mountpoint]._.clients[clientid].status.metaintcycle, chunk.length);
                            mounts[mountpoint]._.clients[clientid].status.metaintcycle = 0;
                            mounts[mountpoint]._.clients[clientid].status.sent += mounts[mountpoint]._.meta.length;
                            mounts[mountpoint]._.bytessent += mounts[mountpoint]._.meta.length;
                        }
                        if (remainchunk.length > 0) {
                            mounts[mountpoint]._.clients[clientid].res.write(remainchunk);
                            mounts[mountpoint]._.clients[clientid].status.metaintcycle = remainchunk.length;
                        }
                    } else {
                        mounts[mountpoint]._.clients[clientid].res.write(chunk);
                        mounts[mountpoint]._.clients[clientid].status.metaintcycle += chunk.length;
                    }
                    mounts[mountpoint]._.clients[clientid].status.sent += chunk.length;
                    mounts[mountpoint]._.bytessent += chunk.length;
                }
            });
            if (mounts[mountpoint].prebuffertime) {
                mounts[mountpoint]._.prebuffers.push(chunk);
                setTimeout(function () {
                    mounts[mountpoint]._.prebuffers.shift();
                }, mounts[mountpoint].prebuffertime);
            }
        });
        res.once('close', function (chunk) {
            setTimeout(function () {
                clienting(mountpoint);
            }, 100);
        });
    }).once('error', function (e) {
        setTimeout(function () {
            clienting(mountpoint);
        }, 100);
        log(mountpoint + '  ' + e);
    });
};

var parseandsetconfig = function (input, callback) {
    input = JSON.parse(input);
    config = input.config;
    mounts = input.mounts;
    if (typeof callback === 'function') {
        callback();
    }
};

var init = function () {
    Object.keys(mounts).forEach(function (mountpoint) {
        clienting(mountpoint);
    });
    
    server.listen(config.port, config.ip);
    
    setInterval(function () {
        Object.keys(mounts).forEach(function (mountpoint) {
            if (mounts[mountpoint]._ &&
                mounts[mountpoint]._.headers['icy-metaint'] &&
                typeof mounts[mountpoint]._.headers['icy-metaint'] === 'number' &&
                mounts[mountpoint]._.headers['icy-metaint'] > 0 &&
                Math.floor(mounts[mountpoint]._.headers['icy-metaint']) === mounts[mountpoint]._.headers['icy-metaint']) {
                
                var metaurl = false;
                if (typeof mounts[mountpoint].metaurl === 'string' && mounts[mountpoint].metaurl.trim() !== '') {
                    metaurl = mounts[mountpoint].metaurl;
                } else if (typeof config.metaurl === 'string' && config.metaurl.trim() !== '') {
                    metaurl = config.metaurl;
                }
                if (metaurl) {
                    var metareqdata = '';
                    http.get(metaurl, function (res) {
                        res.on('data', function (chunk) {
                            metareqdata += chunk;
                        });
                        res.on('end', function () {
                            mounts[mountpoint]._.meta = makemeta(metareqdata);
                        });
                    }).on('error', function (e) {
                        log('getting meta-info for '+mountpoint+' from '+mounts[mountpoint].metaurl+': '+e);
                    });
                }
            }
        });
    }, 1000);
};

var stdin = '';
process.stdin.on('data', function (chunk) {
    stdin += chunk;
});
process.stdin.on('close', function () {
    parseandsetconfig(stdin, function () {
        init();
    });
});